src/System.Private.Uri/src/System/Uri.cs

```diff
@@ -4643,10 +4654,23 @@ private unsafe Check CheckCanonical(char* str, ref ushort idx, ushort end, char
                         res |= Check.DotSlashAttn;
                     }
                 }
-                else if (!needsEscaping && ((c <= '"' && c != '!') || (c >= '[' && c <= '^') || c == '>'
-                    || c == '<' || c == '`'))
+                else if (((c <= '"' && c != '!') || (c >= '[' && c <= '^') || c == '>'
+                        || c == '<' || c == '`'))
                 {
-                    needsEscaping = true;
+                    if (!needsEscaping) needsEscaping = true;
+
+                    // Check above validate only if we have valid Iri characters.
+                    // It's not enough to have valid canonical Iri uri.
+                    // If we have Iri uri(Flags.HasUnicode) we set Check.NotIriCanonical
+                    // also if current part(Path,Query,Fragment) has got valid Iri characters.
+                    // This will lead to invalidate Uri es.
+                    // "http://www.contoso.com/\u00E4/ path2/ param=val" is invalid Iri lead to E_PathNotCanonical
+                    // "http://www.contoso.com/\u00E4? param=val" is invalid Iri lead to E_QueryNotCanonical                    
+                    // "http://www.contoso.com/\u00E4?param=val# fragment" is invalid Iri lead to E_FragmentNotCanonical     
```

Comment:
`// path, query, and fragment will be validated.` Maybe "invalidated". I mean to avoid to "rescan" fragment and check if there are unicode chars if we have HasUnicode this is Iri uri...so if some "part"(path, query, fragment) is invalid we'll have a `NonIriCanonical` that invalidates uri.